src/viem/prepareTransactions.ts

```diff
@@ -0,0 +1,326 @@
+import { TokenBalance, TokenBalanceWithAddress } from 'src/tokens/slice'
+import { TransactionRequestCIP42 } from 'node_modules/viem/_types/chains/celo/types'
+import BigNumber from 'bignumber.js'
+import { Address, encodeFunctionData, EstimateGasExecutionError, Hex, zeroAddress } from 'viem'
+import { estimateFeesPerGas } from 'src/viem/estimateFeesPerGas'
+import { publicClient } from 'src/viem/index'
+import { STATIC_GAS_PADDING } from 'src/config'
+import Logger from 'src/utils/Logger'
+import { Field, SwapTransaction } from 'src/swap/types'
+import erc20 from 'src/abis/IERC20'
+
+// Apply a multiplier for the decreased swap amount to account for the
+// varying gas costs of different swap providers (or even the same swap)
+const DECREASED_SWAP_AMOUNT_GAS_COST_MULTIPLIER = 1.2
+
+interface PreparedTransactionsPossible {
+  type: 'possible'
+  transactions: TransactionRequestCIP42[]
+}
+
+export interface PreparedTransactionsNeedDecreaseSpendAmountForGas {
+  type: 'need-decrease-spend-amount-for-gas'
+  maxGasCost: BigNumber
+  feeCurrency: TokenBalance
+  decreasedSpendAmount: BigNumber
+}
+
+export interface PreparedTransactionsNotEnoughBalanceForGas {
+  type: 'not-enough-balance-for-gas'
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSpendAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+export function getMaxGasCost(txs: TransactionRequestCIP42[]): BigNumber {
+  let maxGasCost = BigInt(0)
+  for (const tx of txs) {
+    if (!tx.gas || !tx.maxFeePerGas) {
+      throw new Error('Missing gas or maxFeePerGas')
+    }
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return new BigNumber(maxGasCost.toString())
+}
+
+export function getFeeCurrencyAddress(feeCurrency: TokenBalance) {
+  return !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined
+}
+
+export async function tryEstimateTransaction({
```

Comment:
mainly the same, besides parameters (symbol and address instead of the whole TokenBalanceWithAddress object)